### DISCUSSION TITLE
[Solved] Should we drop FBX support?

### DISCUSSION THREAD

**Discussion by Jklawreszuk** [stride3d/stride#1738] [General]
To write this text, I was inspired by following website articles *:

1. https://godotengine.org/article/fbx-importer-rewritten-for-godot-3-2-4/
2. https://godotengine.org/fbx-import/
3. https://www.threekit.com/blog/when-should-you-use-fbx-3d-file-format
4. https://en.wikipedia.org/wiki/FBX

*I would like to point out that I am not an expert in computer graphics and I may be wrong here in this text.

First, let's review some information about this format. 
FBX is a popular proprietary file format widely supported by many 3D programs. There are two FBX SDK bindings : for C++ and Python officially supported by Autodesk. It owes its popularity to its wide range of features like animations, bones, skinning etc. Another additional advantage is the FBX speed. Since FBX is a binary format, it loads fast and makes partial updates, so updating the entire file is not necessary.

And here the positive aspects of the described file format comes to an end...

As I pointed out earlier, FBX is a proprietary format and all manipulation of it is managed by closed source SDK know as Autodesk FBX Software Developer Kit. For this reason, as consumers, we don't know exactly how the library works, and it is not possible to reproduce the specification 1 : 1 on the development platform of choice (for example, .NET), so as a result we have to rely on wrappers and making sure there are up-to date.

However, many open-source teams use their own libraries to import FBX, such as Blender, as this would otherwise create a proprietary dependency. Stride Engine is also an open-source project, so it should stay away from closed-source infrastructure. If we take a look at some similar projects, like Godot, creating importer from scratch took over a year just to realize...that it doesn't make sense and droping support immediately. Worth to mention, Stride community is relatively smaller to Godot, so migrating could take even a "little bit" much longer...

When it comes to pactical problems, the current use of native bindings requires an additional c++ compiler to build the entire engine, which is sometimes a little painful and indirectly, makes building Stride code on Linux impossible.

## My personal opinion

I don't think we currently have enough resources to create our own FBX importer, as it would require reverse engineering expertise and a lot of time. Leaving it as is, in my opinion, will continue to block the implementation of the Linux runtime support, which I think is more crucial. There is also currently no good and well-known wrapper available on the market for the reasons I mentioned above. The only wrapper that caught my eye was com.autodesk.fbx, but It is used only for Unity.

In the meantime, many alternatives have appeared to FBX such as glTF or USD, and I would recommend putting in the effort, as they are predictable and many importers are available. 

Removing FBX from the main project, however, would only mean no official support. FBX imports could be provided unofficially through a plugin system in the future (currently not implemented, look at issue here https://github.com/stride3d/stride/issues/1120). Ultimately, I have similar conclusions about the future of FBX as the Godot community, and we should learn lessons from others mistakes

What do you think about ? I would like to know your personal thoughts 😊

**Comment by VaclavElias** [comment]
Not sure if related or sort of related but I am into code only development in Stride and made this working nicely in Windows but it doesn't work on Linux because of Assets library and assets dependencies which I don't know if it is linked to FBX?

The point is when I do code only development lets say I am not interested in any assets, only white box models so there should be option to opt out from Assets processing in the pipeline and so the logic development and compiling could be done also on Linux? 

Pointing here towards plug-in option if it is all related with the above I wrote.

**Comment by NicusorN5** [comment]
No. FBX is the only format that can be exported from Blender that doesn't have broken animations. In most cases, no animation was being imported as all, for example Collada, Stanford, glTF and few more.

Atleast this is the case for 3D models I work with.

**Comment by manio143** [comment]
Removing FBX support would be big hit to Stride because a lot of 3D models available online are in this format and being able to import them without having to re-export it through blender into a different format. Having said that, I do think it should be an optional dependency of the engine. I don't forsee Stride code as a whole to be compilable outside of Windows any time soon, but at least allowing Stride users to compile the engine bits, and definitely to run engine and asset compiler outside of Windows would be a good goal.

Another point to be made is that managed C++/CLI is preventing the Stride integration with FBX to be compiled outside Windows. In my understanding switching to P/Invokes would make the code more portable albeit is a little harder to work with.

**Reply by xen2** [reply]
Totally agree, disabling (temporarily?) FBX with a `#if` directive to unblock the linux/OSX asset toolchain port sounds like a good option to me!

**Reply by NicusorN5** [reply]
Yes, it is possible to use P/Invoke.

**Comment by richard-sim** [comment]
I hate to say it, but this is a classic case of platform developers vs users. Since the vast majority of models and animations available online are in FBX format (and as @NicusorN5 points out, FBX is important for Blender compatibility), I believe that FBX support is a hard requirement for users. Saying FBX support will be restored in the future when a plugin system and then the respective FBX plugin is implemented, in order to ease porting to a new platform where there are no users yet, would cause significant harm to the majority of existing users who do use FBX files.

Perhaps the Godot or Blender FBX libraries can be extracted from their respective repos and ported to a standalone lib? It'd still be a C++ dependency, but that should in no way block ports.

**Reply by VaclavElias** [reply]
I think. The title should be changed in this discussion. After a few comments it is not anymore about dropping it but making an option to opt out in some cases 🙂

**Comment by Eideren** [comment]
Looks like we're pretty much all on the same boat with this discussion, I'll close it off then.

- We're not dropping FBX. 
- The first version of the asset compiler running from a non-windows platform ***may not initially*** have support for FBX.
- No changes for users building their game from windows, targeting windows or linux, or any other platforms.

**Comment by Jklawreszuk** [comment]
@Eideren Yeah, I agree with all the answers for which I thank you 😊 So, closing then.
